storage_controller/src/persistence.rs

```diff
@@ -675,6 +686,130 @@ impl Persistence {
+    #[allow(dead_code)]
+    pub(crate) async fn update_metadata_health_records(
+        &self,
+        metadata_health_records: Vec<MetadataHealthPersistence>,
+    ) -> DatabaseResult<()> {
+        use crate::schema::metadata_health::dsl::*;
+
+        self.with_measured_conn(
+            DatabaseOperation::UpdateMetadataHealth,
+            move |conn| -> DatabaseResult<_> {
+                for update in &metadata_health_records {
+                    diesel::insert_into(metadata_health)
+                        .values(update)
+                        .on_conflict((tenant_id, shard_number, shard_count))
+                        .do_update()
+                        .set((
+                            healthy.eq(update.healthy),
+                            last_scrubbed_at.eq(update.last_scrubbed_at),
+                        ))
+                        .execute(conn)?;
+                }
+                Ok(())
+            },
+        )
+        .await
+    }
+
+    /// Lists all the metadata health records.
+    #[allow(dead_code)]
+    pub(crate) async fn list_metadata_health_records(
+        &self,
+    ) -> DatabaseResult<Vec<MetadataHealthPersistence>> {
+        self.with_measured_conn(
+            DatabaseOperation::ListMetadataHealth,
+            move |conn| -> DatabaseResult<_> {
+                Ok(
+                    crate::schema::metadata_health::table
+                        .load::<MetadataHealthPersistence>(conn)?,
+                )
+            },
+        )
+        .await
+    }
+
+    /// Lists all the metadata health records that is unhealthy.
+    #[allow(dead_code)]
+    pub(crate) async fn list_unhealthy_metadata_health_records(
+        &self,
+    ) -> DatabaseResult<Vec<MetadataHealthPersistence>> {
+        use crate::schema::metadata_health::dsl::*;
+        self.with_measured_conn(
+            DatabaseOperation::ListMetadataHealthUnhealthy,
+            move |conn| -> DatabaseResult<_> {
+                Ok(crate::schema::metadata_health::table
+                    .filter(healthy.eq(false))
+                    .load::<MetadataHealthPersistence>(conn)?)
+            },
+        )
+        .await
+    }
+
+    /// Lists all the metadata health records that have not been updated since an `earlier` time.
+    #[allow(dead_code)]
+    pub(crate) async fn list_outdated_metadata_health_records(
+        &self,
+        earlier: chrono::DateTime<chrono::Utc>,
+    ) -> DatabaseResult<
+        Vec<(
+            String,
+            i32,
+            i32,
+            Option<bool>,
+            Option<chrono::DateTime<chrono::Utc>>,
+        )>,
+    > {
+        use crate::schema::metadata_health;
+        use crate::schema::tenant_shards;
+
+        self.with_measured_conn(
+            DatabaseOperation::ListMetadataHealthOutdated,
+            move |conn| -> DatabaseResult<_> {
+                let query = tenant_shards::table
+                    .left_join(
```

Comment:
What's the point of this join? You already have all the fields you need in the `metadata_health` table.

Reply:
There could be shards that never received an update. Any recommendations to address that case?

Reply:
That makes sense, but it means that newly created shards will show up as unhealthy until they get the first scrub. This can make alerting difficult.

What about inserting a "healthy" entry into `metadata_health` every time we insert into `tenant_shards`? That solves the issue for new tenant shards. As for pre-existing tenant shards, perhaps we can write a migration that inits the `metadata_health` table to mirror `tenant_shards`?

Reply:
Yea I think this approach is good. I will ask for another round of review after adding the insertion logic + migration.

Reply:
c4d082d74821201b3c2a8ca76b45bd0ab8e5960f

- insert into `metadata_health` at tenant shard creation time.
- update migration for setting health status existing tenant shards
- remove left joins; direct filter based on timestamp for listing outdated
- adapt `test_storage_controller_metadata_health`
- remove `MetadataHealthNullable`, no longer needed to keep to optional fields: everything has a value since tenant shard creation.